pkpdapp/pkpdapp/forms/dataset.py

```diff
@@ -15,58 +20,219 @@ def file_size(value):
+                    'Error parsing file, '
+                    '%(filename)s does not have the following columns: '
+                    '%(error_cols)s'
+                )),
+                code='invalid',
+                params={'filename': uploaded_file.name,
+                        'error_cols': error_cols},
+            )
+
+        # check that time unit is only h or d
+        time_units = data['TIME_UNIT'].unique().tolist()
+        error_tunits = []
+        for t_unit in time_units:
+            if t_unit not in ['h', 'd', '.', 'hours']:
+                error_tunits.append(t_unit)
+        if len(error_tunits) > 0:
+            raise forms.ValidationError(
+                _((
+                    'Error parsing file, '
+                    '%(filename)s contains the following unknown time units: '
+                    '%(error_tunits)s'
+                )),
+                code='invalid',
+                params={'filename': uploaded_file.name,
+                        'error_tunits': error_tunits},
+            )
 
-    administration_type = forms.ChoiceField(
-        label='Administration type',
-        help_text='method of drug administration')
+        # check for missing data and drop any rows where data are missing
+        num_missing = data.isna().sum().sum()
+        if num_missing > 0:
+            if not self.instance.id and "warn_missing_data" not in self.data:
+                # creates entry in self.data if a user tries to save again
+                self.add_error('file', format_html(
+                    'Warning! There are ' + str(num_missing) + ' missing data '
+                    'values in file which will be dropped during upload.'
+                    ' To add the dataset anyway, please save again.'
+                    '<input type="hidden" id="warn-missing-data"'
+                    'name="warn_missing_data" value="0"/>')
+                )
+            data = data.dropna()
+        self._data = data
+        return data
 
-    file = forms.FileField(label='Data file', validators=[file_size])
+    def save(self, commit=True):
+        instance = super().save()
+        if self.project_id is not None:
+            project = Project.objects.get(id=self.project_id)
+            project.datasets.add(instance)
+            if commit:
+                project.save()
 
+        # save default biomarker types
+        data = self._data
+        data_without_dose = data.query('DV != "."')
+        bts_unique = data_without_dose[['YDESC', 'UNIT']].drop_duplicates()
+        for index, row in bts_unique.iterrows():
+            if row['UNIT'] == ".":
+                BiomarkerType.objects.create(
+                    name=row['YDESC'],
+                    description="",
+                    dataset=instance)
+            else:
+                unit_query = StandardUnit.objects.filter(symbol=row['UNIT'])
+                if not unit_query:
+                    unit = StandardUnit(symbol=row['UNIT'])
```

Comment:
I don't think we should add to the list of standard units here, as the user could be uploading any any old units. We should restrict possible units in uploaded datafiles to those in the `Unit` table, and convert these to units in the `StandardUnit` table, as is done in the migrations. So all data is stored in standardised units that we control, and can deal with properly when plotting etc.

Reply:
Done.